### PR TITLE
expression: implement vectorized evaluation for `builtinWeekWithModeSig`

### PR DESCRIPTION
**Pull request by Lanearth**
PCP #12103

### What problem does this PR solve? 
implement vectorized evaluation for builtinWeekWithModeSig

### What is changed and how it works?
```
$ make vectorized-bench VB_FILE=Time VB_FUNC=builtinWeekWithModeSig
cd ./expression && \
        go test -v -benchmem \
                -bench=BenchmarkVectorizedBuiltinTimeFunc \
                -run=BenchmarkVectorizedBuiltinTimeFunc \
                -args "builtinWeekWithModeSig"
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinTimeFuncGenerated-8           1000000000               0.00943 ns/op         0 B/op          0 allocs/op
BenchmarkVectorizedBuiltinTimeFunc/builtinWeekWithModeSig-VecBuiltinFunc-8                 28946             41939 ns/op               0 B/op          0 allocs/op
BenchmarkVectorizedBuiltinTimeFunc/builtinWeekWithModeSig-NonVecBuiltinFunc-8              10000            100496 ns/op               0 B/op          0 allocs/op
PASS
ok      github.com/pingcap/tidb/expression      2.765s
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_time_vec.go [hunk 1]
```diff
@@ -890,11 +890,55 @@ func (b *builtinQuarterSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column)
...
+		return err
+	}
+	if err := b.args[1].VecEvalInt(b.ctx, input, buf2); err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf2)
+
+	result.ResizeInt64(n, false)
+	result.MergeNulls(buf1)
+	result.MergeNulls(buf2)
```
Comment (Reminiscent) [inline expression/builtin_time_vec.go]: Here should be ```result.MergeNulls(buf1, buf2)```

Comment (Lanearth) [reply expression/builtin_time_vec.go]: fixed

File: expression/builtin_time_vec.go [hunk 2]
```diff
@@ -887,11 +887,54 @@ func (b *builtinQuarterSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column)
...
+			if err := handleInvalidTimeError(b.ctx, types.ErrIncorrectDatetimeValue.GenWithStackByArgs(date.String())); err != nil {
+				return err
+			}
+			result.SetNull(i, true)
+			continue
+		}
+		mode := int(ms[i])
+		if buf2.IsNull(i) {
+			mode = 0
+		}
```
Comment (wshwsh12) [inline expression/builtin_time_vec.go]: Useless check. If buf2.IsNull(i) is true, the check L919  will be true and continue.

Comment (Lanearth) [reply expression/builtin_time_vec.go]: fixed

File: expression/builtin_time_vec.go [hunk 3]
```diff
@@ -911,11 +911,51 @@ func (b *builtinQuarterSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column)
...
+			continue
+		}
+		date := ds[i]
+		if date.IsZero() {
+			if err := handleInvalidTimeError(b.ctx, types.ErrIncorrectDatetimeValue.GenWithStackByArgs(date.String())); err != nil {
+				return err
+			}
+			result.SetNull(i, true)
+			continue
+		}
```
Comment (wshwsh12) [inline expression/builtin_time_vec.go]: The check order is different from non-vec function's.

1. Check arg0 is null.
2. Check arg0 is zero and handleError .
3. Check arg1 is null.

Comment (Lanearth) [reply expression/builtin_time_vec.go]: fixed

File: expression/builtin_time_vec_test.go [hunk 4]
```diff
@@ -142,6 +142,9 @@ var vecBuiltinTimeCases = map[string][]vecExprBenchCase{
 			},
 		},
 	},
+	ast.Week: {
+		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ETDatetime, types.ETInt}},
```
Comment (wshwsh12) [inline expression/builtin_time_vec_test.go]: The argument arg1 (field mode) is valid when it is 0-7 (more detail can see weekMode function) . I think we should add some test cases for it.

**Pull request merged**